File: sum_primes.exs
# Mariana Ramirez
# A01787819

defmodule PrimeSum do

  def prime?(number) do
    cond do
      number < 2 ->
        false

      number == 2 ->
        true

      rem(number, 2) == 0 ->
        false

      true ->
        limit = :math.sqrt(number) |> trunc()
        check_divisors(number, 3, limit)
    end
  end

  def check_divisors(_number, divisor, limit) when divisor > limit do
    true
  end

  def check_divisors(number, divisor, limit) do
    if rem(number, divisor) == 0 do
      false
    else
      check_divisors(number, divisor + 2, limit)
    end
  end

  def sequential_sum(n) do
    Enum.reduce(2..n, 0, fn number, total ->
      if prime?(number) do
        total + number
      else
        total
      end
    end)
  end

  def sum_interval(first, last) do
    Enum.reduce(first..last, 0, fn number, total ->
      if prime?(number) do
        total + number
      else
        total
      end
    end)
  end

  def create_ranges(n, tasks) do
    size = ceil(n / tasks)

    Enum.map(0..(tasks - 1), fn i ->
      first = i * size + 1
      last = min((i + 1) * size, n)
      {first, last}
    end)
  end

  def parallel_sum(n, tasks \\ System.schedulers_online()) do
    ranges = create_ranges(n, tasks)

    task_list =
      Enum.map(ranges, fn range ->
        Task.async(fn ->
          {first, last} = range
          sum_interval(first, last)
        end)
      end)

    results =
      Enum.map(task_list, fn task ->
        Task.await(task, :infinity)
      end)

    Enum.sum(results)
  end

  def run(n) do
    sequential = sequential_sum(n)
    parallel = parallel_sum(n)

    IO.puts("Sequential result: #{sequential}")
    IO.puts("Parallel result: #{parallel}")
    IO.puts("Same result? #{sequential == parallel}")
  end
end

n =
  case System.argv() do
    [value] -> String.to_integer(value)
    _ -> 1_000_000
  end

PrimeSum.run(n)

File: sum_primes.cpp
/*
Mariana Ramirez
A01787819
 */

#include <iostream>
#include <vector>
#include <thread>
#include <cmath>

using namespace std;

bool isPrime(long n) {
    if (n < 2) {
        return false;
    }

    if (n == 2) {
        return true;
    }

    for (long i = 2; i <= ceil(sqrt(n)); i++) {
        if (n % i == 0) {
            return false;
        }
    }

    return true;
}

unsigned long long sumPrimesSequential(long n) {
    unsigned long long sum = 0;

    for (long i = 2; i <= n; i++) {
        if (isPrime(i)) {
            sum += i;
        }
    }

    return sum;
}

void sumRange(long start, long end, unsigned long long &result) {
    result = 0;

    for (long i = start; i <= end; i++) {
        if (isPrime(i)) {
            result += i;
        }
    }
}

unsigned long long sumPrimesParallel(long n, int numThreads) {
    vector<thread> threads;
    vector<unsigned long long> results(numThreads);

    long rangeSize = n / numThreads;
    long start = 2;

    for (int i = 0; i < numThreads; i++) {
        long end;

        if (i == numThreads - 1) {
            end = n;
        } else {
            end = start + rangeSize - 1;
        }

        threads.push_back(thread(sumRange, start, end, ref(results[i])));

        start = end + 1;
    }

    for (int i = 0; i < numThreads; i++) {
        threads[i].join();
    }

    unsigned long long total = 0;

    for (int i = 0; i < numThreads; i++) {
        total += results[i];
    }

    return total;
}

int main() {
    long n = 1000000;
    int numThreads = thread::hardware_concurrency();

    unsigned long long sequential = sumPrimesSequential(n);
    unsigned long long parallel = sumPrimesParallel(n, numThreads);

    cout << "Sequential result: " << sequential << endl;
    cout << "Parallel result:   " << parallel << endl;

    if (sequential == parallel) {
        cout << "The results are the same." << endl;
    } else {
        cout << "The results are different." << endl;
    }

    return 0;
}